### ISSUE TITLE
Add validation parameters to ConfigItem

### ISSUE THREAD

**Issue by dmlloyd** [quarkusio/quarkus#2489]
Add something like the following to `ConfigItem`:

```java
    String NOT_SPECIFIED = "<<not specified>>";

    String matching() default NOT_SPECIFIED;

    String minValue() default NOT_SPECIFIED;
    String maxValue() default NOT_SPECIFIED;
    boolean minInclusive() default true;
    boolean maxInclusive() default false;
````

The behavior of the `minValue` and `maxValue` parameters should be that for each parameter, if it is given and the value is `Comparable`, then the value should be `convert`ed using the defined converters and compared to the given value, registering an error if it is not within range (which will lead to Quarkus not starting up).  Specifying `minValue` and `maxValue` for values which are not either `Comparable` or `OptionalInt`/`OptionalLong`/`OptionalDouble` would be an error.

The `minInclusive` and `maxInclusive` parameters would define whether the minimum and maximum values are inclusive.

The behavior of `matching` is that the value is compiled to a regular expression `Pattern` which is matched against the given value (which must have a meaningful `toString` method implementation).  Failing to match the pattern would result in an error being registered.

Having both `minValue`/`maxValue` and `matching` would be allowed though probably only useful in a very small set of cases.

**Comment by dmlloyd** [comment]
Additional ideas:

```java
    boolean allowEmpty() default true;
    boolean required() default false;
```

The distinctions between these values and their interactions with `defaultValue` would not be clear until the MP config issues and discussions are resolved:

* eclipse/microprofile-config#407
* smallrye/smallrye-config#83

**Comment by dmlloyd** [comment]
Another idea: print a deprecation warning if the property is populated but is marked `@Deprecated`.  If we wanted to get very fancy we could print out the summary string from the `@deprecated` doc tag.

**Comment by svenhaag** [comment]
Hi, what's the state of this? I would love to see these validation annotations in SmallRye-Config (the MP-Config Spec). Cheers

**Comment by dmlloyd** [comment]
I'm piloting a rough draft of this in https://github.com/smallrye/smallrye-config/pull/310; @radcortez is trying a different approach (reusing Bean Validation annotations) in https://github.com/smallrye/smallrye-config/pull/317.

**Comment by svenhaag** [comment]
Both issues are simply awesome, but I would prefer Bean Validation too over custom annotations. Cheers.

**Comment by geoand** [comment]
This has already been implemented with the introduction of [@ConfigMapping](https://smallrye.io/docs/smallrye-config/mapping/mapping.html) into SmallRye Config 2.3 (which is included in Quarkus 2.0)
